dashboard/modules/job/job_supervisor.py

```diff
@@ -293,159 +950,173 @@ async def run(
-            raise RuntimeError(
-                f"Job {self._job_id} is already in RUNNING state. "
-                f"JobSupervisor.run() should only be called once. "
-            )
-        if curr_status != JobStatus.PENDING:
-            raise RuntimeError(
-                f"Job {self._job_id} is not in PENDING state. "
-                f"Current status is {curr_status} with message {curr_message}."
-            )
-
-        if _start_signal_actor:
-            # Block in PENDING state until start signal received.
-            await _start_signal_actor.wait.remote()
-
-        driver_agent_http_address = (
-            "http://"
-            f"{ray.worker.global_worker.node.node_ip_address}:"
-            f"{ray.worker.global_worker.node.dashboard_agent_listen_port}"
-        )
-        driver_node_id = ray.worker.global_worker.current_node_id.hex()
-
-        await self._job_info_client.put_status(
-            self._job_id,
-            JobStatus.RUNNING,
-            jobinfo_replace_kwargs={
-                "driver_agent_http_address": driver_agent_http_address,
-                "driver_node_id": driver_node_id,
-            },
-        )
 
         try:
             # Configure environment variables for the child process. These
             # will *not* be set in the runtime_env, so they apply to the driver
             # only, not its tasks & actors.
-            os.environ.update(self._get_driver_env_vars(resources_specified))
-            self._logger.info(
-                "Submitting job with RAY_ADDRESS = "
-                f"{os.environ[ray_constants.RAY_ADDRESS_ENVIRONMENT_VARIABLE]}"
-            )
+            os.environ.update(self._get_driver_env_vars())
+
+            self._logger.info(f"({self._job_id}) Executing job driver's entrypoint")
+
             log_path = self._log_client.get_log_file_path(self._job_id)
-            child_process = self._exec_entrypoint(log_path)
-            child_pid = child_process.pid
 
+            # Execute job's entrypoint in the subprocess
+            self._driver_process = self._exec_entrypoint(log_path)
+
+        except Exception as e:
+            self._logger.error(
+                f"({self._job_id}) Got unexpected exception while executing job's entrypoint: {repr(e)}",
+                exc_info=e,
+            )
+
+    async def join(self) -> JobExecutionResult:
+        """
+        Joins job execution blocking until either of the following conditions is true
+
+        1. Job driver has completed (subprocess exited)
+        2. Job has been interrupted (due to Ray job being stopped)
+
+        NOTE: This method should only be called after `JobExecutor.start`
+        """
+
+        try:
+            assert self._driver_process, "Job's driver is not running"
+
+            child_process = self._driver_process
+
+            # Block until either of the following occurs:
+            #   - Process executing job's entrypoint completes (exits, returning specific exit-code)
```

Comment:
Changes

 - Splitting job running sequence in 2 methods (so that job is annotated as RUNNING only if we're able to successfully start the entrypoint)
 - No logic changes otherwise)